lucene/test-framework/src/java/org/apache/lucene/util/VerifyTestClassNamingConvention.java

```diff
@@ -0,0 +1,98 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.util;
+
+import com.carrotsearch.randomizedtesting.RandomizedContext;
+import org.junit.Assume;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * Enforce test naming convention.
+ */
+public class VerifyTestClassNamingConvention extends AbstractBeforeAfterRule {
+  public static final Pattern ALLOWED_CONVENTION = Pattern.compile("(.+?)\\.Test[^.]+");
+
+  private static Set<String> exceptions;
+  static {
+    try {
+      exceptions = new HashSet<>();
+      try (BufferedReader is =
+             new BufferedReader(
+                 new InputStreamReader(
+                   VerifyTestClassNamingConvention.class.getResourceAsStream("test-naming-exceptions.txt"),
+                   StandardCharsets.UTF_8))) {
+        is.lines().forEach(exceptions::add);
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  protected void before() throws Exception {
+    if (TestRuleIgnoreTestSuites.isRunningNested()) {
+      // Ignore nested test suites that test the test framework itself.
+      return;
+    }
+
+    String suiteName = RandomizedContext.current().getTargetClass().getName();
+
+    // You can use this helper method to dump all suite names to a file.
+    // Run gradle with one worker so that it doesn't try to append to the same
+    // file from multiple processes:
+    //
+    // gradlew  test --max-workers 1 -Dtests.useSecurityManager=false
+    //
+    // dumpSuiteNamesOnly(suiteName);
+
+    if (!ALLOWED_CONVENTION.matcher(suiteName).matches()) {
+      // if this class exists on the exception list, leave it.
+      if (!exceptions.contains("!" + suiteName)) {
+        throw new AssertionError("Suite must follow Test*.java naming convention: "
+          + suiteName);
+      }
+    }
+  }
+
+  private void dumpSuiteNamesOnly(String suiteName) throws IOException {
+    // Has to be a global unique path (not a temp file because temp files
+    // are different for each JVM).
+    Path temporaryFile = Paths.get("c:\\_tmp\\test-naming-exceptions.txt");
```

Comment:
Hmm this is effectively a `// nocommit` right?  I.e. we must find the right place for this to live in the sources?

Reply:
Mike, this patch was just to show how it can be done. And this method only collects test cases for exclusion - it's not used and could be removed since in theory we'd only want to take away from once-generated exceptions file. I left it to show how I collected the list in the first place.